src/main/java/net/spy/memcached/tls/TLSConnectionManager.java

```diff
@@ -0,0 +1,240 @@
+        this.sslContext = sslContext;
+    }
+
+    private void initSSLEngine() {
+        // We are the client, not the server
+        if (sslEngine != null) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Resetting SSL Engine");
+            }
+            closeSslEngine();
+        }
+        sslEngine = sslContext.createSSLEngine();
+        sslEngine.setUseClientMode(true);
+    }
+
+    private void initBuffers(SSLSession session) {
+        // allocateDirect() to keep all bytes contiguous in memory
+        appOutBuffer = ByteBuffer.allocateDirect(session.getApplicationBufferSize());
+        appInBuffer = ByteBuffer.allocateDirect(session.getApplicationBufferSize());
+        networkOutBuffer = ByteBuffer.allocateDirect(session.getPacketBufferSize());
+        networkInBuffer = ByteBuffer.allocateDirect(session.getPacketBufferSize());
+    }
+
+    public boolean doHandshake(SocketChannel socketChannel) throws IOException {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("%s - Beginning handshake.", socketChannel.getRemoteAddress());
+        }
+
+        try {
+            initSSLEngine();
+            sslEngine.beginHandshake();
+            currentSession = sslEngine.getSession();
+            initBuffers(currentSession);
+
+            HandshakeStatus handshakeStatus = sslEngine.getHandshakeStatus();
+            if (LOG.isDebugEnabled()) {
+                LOG.info("%s - Handshake status: %s", socketChannel.getRemoteAddress(), handshakeStatus.name());
+            }
+            while (!Thread.currentThread().isInterrupted()
+                    && handshakeStatus != HandshakeStatus.FINISHED
+                    && handshakeStatus != HandshakeStatus.NOT_HANDSHAKING) {
+
+                switch (handshakeStatus) {
+                    case NEED_TASK:
+                        // we need to finish these tasks for the handshake to continue
+                        List<Future<?>> tasks = new ArrayList<>();
+                        Runnable currentTask = sslEngine.getDelegatedTask();
+                        while(currentTask != null) {
+                            tasks.add(executor.submit(currentTask));
+                            currentTask = sslEngine.getDelegatedTask();
+                        }
+                        for (Future<?> task : tasks) {
+                            try {
+                                task.get();
+                            } catch (ExecutionException e) {
+                                // TODO make this more clear
+                                throw new RuntimeException(e);
+                            }
+                        }
+                        break;
+                    case NEED_WRAP:
+                        // We're about to send something to the server, but it needs to be wrapped first
+                        networkOutBuffer.clear();
+                        SSLEngineResult wrapResult = sslEngine.wrap(appOutBuffer, networkOutBuffer);
```

Comment:
I'm not sure that we need to handle application data at all in the handshake, everything I've seen shows this coming later for communication post handshake

Reply:
As far as I can tell, the name "appOutBuffer" here is in line with the methods that get called. This buffer holds the data that makes up the individual messages we're sending to the server as part of the handshake.